pkg/standalone-utils/contracts/WeightedLPOracle.sol

```diff
@@ -0,0 +1,280 @@
+import { IERC20 } from "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+
+import { IWeightedPool } from "@balancer-labs/v3-interfaces/contracts/pool-weighted/IWeightedPool.sol";
+import { Rounding } from "@balancer-labs/v3-interfaces/contracts/vault/VaultTypes.sol";
+import { IVault } from "@balancer-labs/v3-interfaces/contracts/vault/IVault.sol";
+
+import { InputHelpers } from "@balancer-labs/v3-solidity-utils/contracts/helpers/InputHelpers.sol";
+import { FixedPoint } from "@balancer-labs/v3-solidity-utils/contracts/math/FixedPoint.sol";
+
+contract WeightedLPOracle is AggregatorV3Interface {
+    using FixedPoint for uint256;
+    using SafeCast for *;
+
+    error UnsupportedDecimals();
+
+    uint256 internal constant _WAD_DECIMALS = 18;
+
+    IVault internal immutable _vault;
+    IWeightedPool public immutable pool;
+    uint256 internal immutable _version;
+    uint256 internal immutable _totalTokens;
+    string internal _description;
+
+    uint256 internal immutable _weight0;
+    uint256 internal immutable _weight1;
+    uint256 internal immutable _weight2;
+    uint256 internal immutable _weight3;
+    uint256 internal immutable _weight4;
+    uint256 internal immutable _weight5;
+    uint256 internal immutable _weight6;
+    uint256 internal immutable _weight7;
+
+    AggregatorV3Interface internal immutable _feedToken0;
+    AggregatorV3Interface internal immutable _feedToken1;
+    AggregatorV3Interface internal immutable _feedToken2;
+    AggregatorV3Interface internal immutable _feedToken3;
+    AggregatorV3Interface internal immutable _feedToken4;
+    AggregatorV3Interface internal immutable _feedToken5;
+    AggregatorV3Interface internal immutable _feedToken6;
+    AggregatorV3Interface internal immutable _feedToken7;
+
+    uint256 internal immutable _feedToken0DecimalScalingFactor;
+    uint256 internal immutable _feedToken1DecimalScalingFactor;
+    uint256 internal immutable _feedToken2DecimalScalingFactor;
+    uint256 internal immutable _feedToken3DecimalScalingFactor;
+    uint256 internal immutable _feedToken4DecimalScalingFactor;
+    uint256 internal immutable _feedToken5DecimalScalingFactor;
+    uint256 internal immutable _feedToken6DecimalScalingFactor;
+    uint256 internal immutable _feedToken7DecimalScalingFactor;
+
+    constructor(IVault vault_, IWeightedPool pool_, AggregatorV3Interface[] memory feeds, uint256 version_) {
+        _version = version_;
+        _vault = vault_;
+        pool = pool_;
+
+        IERC20[] memory tokens = vault_.getPoolTokens(address(pool_));
+        uint256[] memory weights = pool_.getNormalizedWeights();
+        uint totalTokens = tokens.length;
+
+        _totalTokens = totalTokens;
+        _description = string.concat(IERC20Metadata(address(pool)).symbol(), "/USD");
+
+        InputHelpers.ensureInputLengthMatch(totalTokens, feeds.length);
+
+        for (uint256 i = 0; i < totalTokens; i++) {
```

Comment:
well technically we don't need a for loop here; we can just check the length and do `else / if` until the end. Doesn't make much of a difference though.